### ISSUE TITLE
Incorrect serialization of collections.Counter objects.

### ISSUE THREAD

**Issue by robertnishihara** [ray-project/ray#3846]
```python
from collections import Counter
import ray

ray.init(num_cpus=0)

c = Counter([1, 2, 3])
print(c)  # Counter({1: 1, 2: 1, 3: 1})

new_c = ray.get(ray.put(c))
print(new_c)  # Counter()
```

**Comment by richardliaw** [comment]
Maybe when this is addressed, there should be a test suite for all collections objects because defaultdict has also been problematic in the past.

**Comment by robertnishihara** [comment]
Fixed now, by the arrow update and tested in #3982.
